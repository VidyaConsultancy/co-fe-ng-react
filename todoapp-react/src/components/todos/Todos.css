.app-todos {
  display: grid;
  padding: 2em;
  width: min(500px, 60%);
  margin-inline: auto;
}

.todos-header {
  padding: 0 1em 0.5em;
  background-color: #fff;
  border-bottom: 3px double #dbdbdb;
}

.todos {
  background-color: #fffff8;
  margin: 0;
  list-style: none;
  padding: 0;
}

.todos-item {
  padding: 0.5em 1em;
  border-bottom: 1px solid #dbdbdb;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.todo-toggle {
  flex: 1;
  cursor: pointer;
  display: inline-flex;
  align-items: center;
  column-gap: 0.5em;
}

.todos-title {
  text-align: center;
  font-size: 2rem;
  font-family: monospace;
  padding: 0.25em 1em;
}

.todos-footer {
  padding: 2em 1em;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #fff;
  box-shadow:
    0 3px 0 rgba(0, 0, 0, 0.3),
    0 6px 0 rgba(0, 0, 0, 0.10),
    0 9px 0 rgba(0, 0, 0, 0.05);
}

.app-todo {
  padding: 2em 2em;
}

.todo-detail {
  width: 100%;
  max-width: 500px;
  margin-inline: auto;
}

.button-wrapper {
  padding-block: 1em;
  margin-bottom: 2em;
}
