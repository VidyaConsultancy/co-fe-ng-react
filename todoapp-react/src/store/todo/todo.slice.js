import { createSlice } from '@reduxjs/toolkit';

const todoSlice = createSlice({
  name: 'todo',
  initialState: {
    todos: [],
    isFetching: false,
    error: null,
  },
  reducers: {
    getAllTodo: (state, action) => {
      state.todos = action.payload.todos;
    }
  }
})

export const { getAllTodo } = todoSlice.actions;
export const todoReducer = todoSlice.reducer;

/* const reducer = (state = { isFetching: false }, action) => {
  if(action.type === 'something') {
    return { ...state, todos: action.payload.todos }
  }
} */
