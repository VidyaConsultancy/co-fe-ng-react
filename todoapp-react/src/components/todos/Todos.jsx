import React from "react";
import { useDispatch, useSelector, connect } from "react-redux";
import {
  fetchAllTodos,
  addNewTodo,
  updateATodo,
  deleteATodo,
} from "../../store/todo/todo.slice";
import { TodoFooter } from "../todo-footer/TodoFooter";
import { TodoHeader } from "../todo-header/TodoHeader";
import { TodoList } from "../todo-list/TodoList";
import "./Todos.css";

// react hooks introduced in 16.8v
// useState, useEffect, useCallback, useLayoutEffect, useReducer, useRef, useId

export const Todos = () => {
  const dispatch = useDispatch();
  const todoState = useSelector((state) => state.todo);

  const handleAddTodo = async (todoTitle) => {
    if (todoTitle.trim().length === 0) return false;
    const newTodo = { todo: todoTitle, isCompleted: false };
    dispatch(addNewTodo(newTodo));
  };

  const handleTodoDelete = async (todoId) => {
    dispatch(deleteATodo(todoId));
  };

  const handleUpdateTodo = async (updatedTodo) => {
    dispatch(updateATodo(updatedTodo));
  };

  /* const fetchTodos = () => {
    fetch(`http://localhost:4000/todos`, {
      method: 'GET'
    }).then(res => {
      return res.json(); 
    }).then(data => {
      console.log(data);
    }).catch(error => console.error(error));
  } */

  React.useEffect(() => {
    dispatch(fetchAllTodos());
  }, [dispatch]);

  return (
    <div className="container app-todos" data-testid="todosContainer">
      {todoState.error ? <small>{todoState.error.message}</small> : null}
      <TodoHeader handleSubmit={handleAddTodo} />
      <TodoList
        todos={todoState.todos}
        handleTodoDelete={handleTodoDelete}
        handleUpdateTodo={handleUpdateTodo}
      />
      <TodoFooter
        total={todoState.todos.length}
        completed={todoState.completed}
      />
    </div>
  );
};

/**
 * Phase I Mounting -> invoked only once
 * constructor
 * static getDerivedStateFromProps
 * render
 * componentDidMount
 * Phase II Updating
 * static getDerivedStateFromProps
 * shouldComponentUpdate
 * render
 * static getSnapshotBeforeUpdate
 * componentDidUpdate
 * Phase III Unmounting ->  invoked only once
 * componentWillUnmount
 */
export class TodosComponent extends React.Component {
  constructor(props) {
    super(props);
    this.appTitle = "Todos";
    this.state = {
      title: "Todos",
      todos: ["Learn HTML"],
    };
    // this.handleClick = this.handleClick.bind(this);
  }

  componentDidMount() {
    this.props.fetchTodos();
  }

  handleClick = () => {
    this.setState({ title: "Todos updated" });
  };

  render() {
    return (
      <div className="container">
        <h2 title={this.state.title}>
          {this.state.title} || {this.appTitle}
        </h2>
        <button onClick={this.handleClick}>Click Me</button>
        <TodoList
          todos={this.props.todos}
        />
      </div>
    );
  }
}
const mapStateToProps = (state) => ({
  todos: state.todo.todos,
  isFetching: state.todo.isFetching,
  accessToken: state.auth.accessToken,
})

const mapDispatchToProps = (dispatch) => ({
  addNewTodo: (todo) => dispatch(addNewTodo(todo)),
  fetchTodos: () => dispatch(fetchAllTodos())
})
export const TodosComponentContainer = connect(mapStateToProps, mapDispatchToProps)(TodosComponent);

