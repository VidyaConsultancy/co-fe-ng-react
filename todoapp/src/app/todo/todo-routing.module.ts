import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { TodoDetailComponent } from './todo-detail/todo-detail.component';

import { TodosComponent } from './todos/todos.component';

const routes: Routes = [
  {
    path: '',
    component: TodosComponent,
  },
  {
    path: ':id',
    component: TodoDetailComponent
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class TodoRoutingModule {}
