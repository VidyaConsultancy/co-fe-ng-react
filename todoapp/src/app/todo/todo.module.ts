import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';

import { TodosComponent } from './todos/todos.component';

@NgModule({
  declarations: [TodosComponent],
  imports: [CommonModule, FormsModule],
  exports: [TodosComponent],
})
export class TodoModule {}
