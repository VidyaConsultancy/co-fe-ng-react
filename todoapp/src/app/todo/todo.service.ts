import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { Todo } from './models/todo.class';

@Injectable({
  providedIn: 'root', // 'root', 'platform', 'any'
})
export class TodoService {
  ops: string[] = [];
  
  constructor(private http: HttpClient) {}

  createTodo(todo: string) {
    const todoObj = new Todo(todo);
    return this.http.post<Todo>(`http://localhost:3000/todos`, todoObj);
  }

  getTodos(): Observable<Todo[]> {
    return this.http.get<Todo[]>(`http://localhost:3000/todos`);
  }

  getTodoById(id: number) {
    return this.http.get<Todo>(`http://localhost:3000/todos/${id}`);
  }

  deleteTodo(id: number) {
    return this.http.delete(`http://localhost:3000/todos/${id}`)
  }

  updateTodo(todo: Todo) {
    return this.http.put<Todo>(`http://localhost:3000/todos/${todo.id}`, todo);
  }

  add(x: number, y: number) {
    this.ops.push('add');
    return x + y;
  }
}
