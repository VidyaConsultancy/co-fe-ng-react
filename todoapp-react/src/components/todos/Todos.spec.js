import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { Todos } from "./Todos";
import { todoReducer } from '../../store/todo/todo.slice';

describe('Todos', () => {
  const store = configureStore({ reducer: {
    todo: todoReducer
  }});

  it('should render component', () => {
    const view = render(<Provider store={store}>
      <Todos />
    </Provider>)
    expect(view.baseElement).toMatchSnapshot();
  })
})
