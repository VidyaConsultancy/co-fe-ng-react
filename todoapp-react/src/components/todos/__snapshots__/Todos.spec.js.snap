// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos should render component 1`] = `
<body>
  <div>
    <div
      class="container app-todos"
      data-testid="todosContainer"
    >
      <div
        class="todos-header"
      >
        <h2
          class="todos-title"
          title="Todos"
        >
          Todos
        </h2>
        <form
          data-testid="addTodoForm"
        >
          <div
            class="form-flex"
          >
            <div
              class="form-group"
            >
              <label
                class="sr-only"
                for="todo"
              >
                What would you like to do?
              </label>
              <input
                class="form-input"
                id="todo"
                name="todo"
                placeholder="What would you like to do?"
                type="text"
                value=""
              />
            </div>
            <button
              class="button button-default"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
      <ul
        class="todos"
      >
        <div
          class="todo-zero-state"
        >
          You don't have any todos yet. Create one here
        </div>
      </ul>
      <div
        class="todos-footer"
      />
    </div>
  </div>
</body>
`;
